### ISSUE TITLE
ANY clause does not generate correct SQL for forced-type fields

### ISSUE THREAD

**Issue by maxsumrall** [jOOQ/jOOQ#10695]
### Expected behavior
Generate an any clause with an array of the column type.

### Actual behavior
Generates an any clause with an array of the column type with huge length. 

### Steps to reproduce the problem

Have a table with a column that is a forced type. Configured in the pom.xml like:
```
<forcedType>
                 <userType>tech.picnic.api.Status</userType>
                 <enumConverter>true</enumConverter>
                  <includeExpression>.*\.order.status</includeExpression>
</forcedType>
```
Write a query using the any clause:
```
context
.select()
.from(ORDER)
where(ORDER.STATUS.eq(DSL.any(List.of(PENDING).toArray(Status[]::new)))
.limit(1)
.fetch()
```
The generate SQL:
```
...omitted...
where ORDER.STATUS = any (?::varchar(2147483647)[])
```

Resulting error: 
```
 ERROR: length for type varchar cannot exceed 10485760
```

### Versions

- jOOQ: 3.13.2
- Java: 11
- Database (include vendor): PG 11
- OS: OSX
- JDBC Driver (include name if inofficial driver): ?

**Comment by lukaseder** [comment]
Thanks for your report.

Can you share the `CREATE TABLE` statement of `order_details`, and possibly the type definition of `order_details.status`, if any?

**Comment by maxsumrall** [comment]
I've created a simple example that:


```
CREATE TABLE jooq
(
    id BIGSERIAL PRIMARY KEY,
    status VARCHAR NOT NULL
)
```

Forced type for the enum converter:
```
<forcedType>
                                    <userType>tech.picnic.finance.token.data.service.Status</userType>
                                    <enumConverter>true</enumConverter>
                                    <includeExpression>.*\.jooq.status</includeExpression>
                                </forcedType>
```

Status: 
```
public enum Status {
    COOL,
    LAME
}
```

Failing example
```
context.insertInto(JOOQ).columns(JOOQ.STATUS).values(COOL).execute();
context.select().from(JOOQ).where(JOOQ.STATUS.eq(DSL.any(COOL, LAME))).fetch();
```



Example stacktrace:
```
org.springframework.dao.DataIntegrityViolationException: jOOQ; SQL [select JOOQ.ID, JOOQ.STATUS from JOOQ where JOOQ.STATUS = any (?::varchar(2147483647)[])]; ERROR: length for type varchar cannot exceed 10485760
  Position: 68; nested exception is org.postgresql.util.PSQLException: ERROR: length for type varchar cannot exceed 10485760
  Position: 68

	at org.jooq_3.14.4.POSTGRES.debug(Unknown Source)
	at org.springframework.jdbc.support.SQLStateSQLExceptionTranslator.doTranslate(SQLStateSQLExceptionTranslator.java:104)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:72)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81)
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:81)
	at com.picnic.jooq.JooqExceptionTranslator.translate(JooqExceptionTranslator.java:72)
	at com.picnic.jooq.JooqExceptionTranslator.handle(JooqExceptionTranslator.java:62)
	at com.picnic.jooq.JooqExceptionTranslator.exception(JooqExceptionTranslator.java:34)
	at org.jooq.impl.ExecuteListeners.exception(ExecuteListeners.java:274)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:390)
	at org.jooq.impl.AbstractResultQuery.fetch(AbstractResultQuery.java:333)
	at org.jooq.impl.SelectImpl.fetch(SelectImpl.java:2880)
	at tech.picnic.finance.token.data.repository.TokenRepositoryTest.aap(TokenRepositoryTest.java:125)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:688)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:210)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:206)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:131)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:65)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:143)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:129)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:127)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:126)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:84)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:71)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)
Caused by: org.postgresql.util.PSQLException: ERROR: length for type varchar cannot exceed 10485760
  Position: 68
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2553)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2285)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:323)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:473)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:393)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:164)
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:153)
	at org.jooq.tools.jdbc.DefaultPreparedStatement.execute(DefaultPreparedStatement.java:214)
	at org.jooq.impl.Tools.executeStatementAndGetFirstResultSet(Tools.java:4192)
	at org.jooq.impl.AbstractResultQuery.execute(AbstractResultQuery.java:279)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:375)
	... 68 more
```

**Comment by maxsumrall** [comment]
Of course it works fine when you get rid of the forced type, e.g. 
```
        context.select().from(JOOQ).where(JOOQ.STATUS.cast(String.class).eq(DSL.any(COOL.toString()))).fetch();
```
works without issue.

**Comment by maxsumrall** [comment]
On closer inspection, the problem occurs only for `VARCHAR` columns. Columns of type `VARCHAR(n)` and `TEXT` column (I'm using PG), don't have any issues. It's just for unbounded `VARCHAR` that it tries to create a max_int array which produces the above error.

**Comment by lukaseder** [comment]
Thanks for following up and for sharing those details.

I think this is still a bug. I can reproduce the problem more simply:

```sql
select cast('a' as varchar(2147483647));
```

But when I write any of these with jOOQ:

```java
ctx.select(cast("a", VARCHAR)).fetch();
ctx.select(cast(new String[] { "a" }, VARCHAR.getArrayDataType())).fetch();
```

Then I get correct SQL:

```sql
-- We might discuss the unnecessary double cast, but that's not the topic here:
select cast('a' as varchar)
select cast(cast('{"a"}' as varchar[]) as varchar[])
```

If `10485760` is a special maximum varchar length in PostgreSQL, we should generate that, or no upper bound at all, and not `Integer.MAX_VALUE`

**Comment by lukaseder** [comment]
This doesn't reproduce it:

```java
DataType<Status> t = VARCHAR.asConvertedDataType(new EnumConverter<>(String.class, Status.class));

Result<Record1<Boolean>> result = ctx.select(field(inline(Status.A, t).eq(any(Status.A, Status.B)))).fetch();
System.out.println(result);
```

It generates the expected SQL:

```sql
select ('A' = any (?::varchar[]))
```

**Comment by lukaseder** [comment]
I've tried this with generated code as well, but couldn't reproduce it either. I'm getting the expected SQL:

```sql
select "public"."t_10695"."e" from "public"."t_10695" where "public"."t_10695"."e" = any (?::varchar[])
```

With this table:

```sql
create table t_10695 (
  e varchar
)
```

This code gen config:

```xml
<forcedType>
    <enumConverter>true</enumConverter>
    <userType>org.jooq.test.postgres.types.E10695</userType>
    <includeExpression>(?i:(.*?\.)T_10695\.E)</includeExpression>
</forcedType>
```

And this test:

```java
@Test
public void testPostgresEnumConverterOnVarcharTypes() {
    clean(T_10695);

    assertEquals(1, create()
        .insertInto(T_10695)
        .columns(T_10695.E)
        .values((E10695) null)
        .execute());

    assertEquals(1, create()
        .insertInto(T_10695)
        .columns(T_10695.E)
        .values(E10695.A)
        .execute());

    Result<?> r1 = create().selectFrom(T_10695).orderBy(T_10695.E.asc().nullsFirst()).fetch();
    assertEquals(2, r1.size());
    assertEquals(asList(null, E10695.A), r1.getValues(T_10695.E));

    T_10695Record r2 = create().selectFrom(T_10695).where(T_10695.E.eq(any(E10695.A, E10695.B))).fetchOne();
    assertEquals(E10695.A, r2.getE());
}
```

**Comment by lukaseder** [comment]
I see you're using 3.13. There have been a few improvements to the `DataType` implementation. Perhaps this got fixed in 3.14...

**Comment by maxsumrall** [comment]
Ah, I have upgraded to 3.14.4 in my comments from this week, where I was able to reproduce this in `VARCHAR`...

**Comment by lukaseder** [comment]
So, you cannot reproduce this with 3.14.4 anymore, either?

**Comment by maxsumrall** [comment]
Sorry for the confusion, I double checked and indeed I am reproducing this on 3.14.4. (I also noticed it's in my stacktrace above https://github.com/jOOQ/jOOQ/issues/10695#issuecomment-737352504)

**Comment by lukaseder** [comment]
So, what's the difference between your attempt and mine? :) Do you have time to create an MCVE? https://github.com/jOOQ/jOOQ-mcve

**Comment by maxsumrall** [comment]
> So, what's the difference between your attempt and mine? :) Do you have time to create an MCVE? https://github.com/jOOQ/jOOQ-mcve

Yeah, it looks like our examples are the same :( 
Will see if I can create a reproducible MCVE or if when creating I figure out something else. :)

**Comment by lukaseder** [comment]
I can try again on 3.14.5-SNAPSHOT. I believe I tried this on 3.15.0-SNAPSHOT, yesterday.

**Comment by lukaseder** [comment]
Nope, the test still passes on 3.14.5-SNAPSHOT

**Comment by maxsumrall** [comment]
https://github.com/maxsumrall/jOOQ-mcve/commit/63091a7c8119ec345a3fef4ef5737eb1944476d5

The issue appears to be caused by generating the jOOQ code using the DDL Database (H2). When I switch to spinning up a Postgres DB and generating the SQL that way, the issue is resolved. I've replicated this in the above MCVE. Particularly, this seems to be where the issue arises, and is the difference between generating the jOOQ code with DDL Database and Postgres: https://github.com/maxsumrall/jOOQ-mcve/commit/63091a7c8119ec345a3fef4ef5737eb1944476d5#diff-748b005cda13a6dcaeef2d3f775f76c61d44ffe69b8e926a8f968e385e1e4fc4R60


DDL/H2:
```
public final TableField<JooqRecord, Status> STATUS = createField(DSL.name("STATUS"), SQLDataType.VARCHAR(2147483647), this, "", new EnumConverter<String, Status>(String.class, Status.class));
```

Postgres:
```
    public final TableField<JooqRecord, Status> STATUS = createField(DSL.name("status"), SQLDataType.VARCHAR, this, "", new EnumConverter<String, Status>(String.class, Status.class));
```

**Comment by maxsumrall** [comment]
The above is simply because (afaik) H2 does not support unlimited length VARCHAR, so uses this `2147483647` value.

**Comment by maxsumrall** [comment]
Simply, H2 supports wider varchar types than postgres. But I'm surprised this wouldn't cause problems elsewhere...

**Comment by katzyn** [comment]
For H2 `CHARACTER VARYING` / `VARCHAR` without a specified length (length is mandatory by the Standard, but not by H2) should be generated when length isn't known; such definition is assumed by all versions of H2 as a character string with maximum supported length. The upcoming H2 2.0 will reject explicit `VARCHAR(2147483647)`, very long values are really supported only by LOB data types.

Actually you can't even put 2147483647 characters into `java.lang.String`.

**Comment by maxsumrall** [comment]
I've found a different but related issue:

Given two tables with a column `STATUS`, where in table `A` this column is of type `VARCHAR` and in table `B` it's of type `VARCHAR(100)`, and a `forced type` to an enum value that applies to both columns, when doing a query like the above, 

```
        context.select().from(A).where(A.STATUS.eq(DSL.any(Status.COOL))); 
```

jOOQ will bind the enum Status to potentially either the `VARCHAR` or `VARCHAR(100)`, regardless that this is a query against `A.STATUS` and therefor should only bind to `VARCHAR`.

I've hit this exact bug in production wondering how this happens, since I was sure I could use the `any` clause safely for this particular column since it was `VARCHAR(N)` which I established earlier worked correctly. However, I was still getting the error `length for type varchar cannot exceed 10485760` due to this situation. Even more unfortunately, this only seems to happen in my production environment but not in a local or dev environment. I was however able to replicate it in my MCVE.
